src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -7191,332 +7218,316 @@ def get_service_connector_type(
-                # Create a new service connector
-                else:
-                    connector_name = full_stack.name
-                    connector_config = connector_id_or_info.configuration
-                    connector_config[
-                        "generate_temporary_tokens"
-                    ] = not need_to_generate_permanent_tokens
+                orchestrator_components = stack.components[
+                    StackComponentType.ORCHESTRATOR
+                ]
+                for orchestrator_component in orchestrator_components:
+                    if isinstance(orchestrator_component, UUID):
+                        orchestrator = self.get_stack_component(
+                            orchestrator_component,
+                            hydrate=False,
+                        )
+                        need_to_generate_permanent_tokens = (
+                            orchestrator.flavor.startswith("vm_")
+                        )
+                    else:
+                        need_to_generate_permanent_tokens = (
+                            orchestrator_component.flavor.startswith("vm_")
+                        )
 
-                    while True:
-                        try:
-                            service_connector_request = ServiceConnectorRequest(
-                                name=connector_name,
-                                connector_type=connector_id_or_info.type,
-                                auth_method=connector_id_or_info.auth_method,
-                                configuration=connector_config,
-                                user=full_stack.user,
-                                workspace=full_stack.workspace,
-                                labels={
-                                    k: str(v)
-                                    for k, v in full_stack.labels.items()
-                                },
-                            )
-                            service_connector_response = (
-                                self.create_service_connector(
+                for connector_id_or_info in stack.service_connectors:
+                    # Fetch an existing service connector
+                    if isinstance(connector_id_or_info, UUID):
+                        existing_service_connector = (
+                            self.get_service_connector(connector_id_or_info)
+                        )
+                        if need_to_generate_permanent_tokens:
+                            if (
+                                existing_service_connector.configuration.get(
+                                    "generate_temporary_tokens", None
+                                )
+                                is not False
+                            ):
+                                connector_config = (
+                                    existing_service_connector.configuration
+                                )
+                                connector_config[
```

Comment:
I know this is not a change of this PR, but do I understand this correctly: If someone creates a stack now that has any vm orchestrator in them, we secretly update the service connector to generate permanent tokens? That doesn't seem right on many levels, especially having such a flavor specific logic in the general implementation of this function?

Reply:
@stefannica @avishniakov

Reply:
@schustmi is not wrong, this feels very specific and doesn't belong here. I guess this was required to make it easier to implement the stack creation wizard in the UI. Perhaps we can move this logic in the UI as well.

The design-wise correct solution for this is to make this behavior dynamic, so we don't have to configure it at all. The skypilot orchestrator would simply ask the service connector explicitly to not generate temporary tokens. Perhaps I can pitch in this part of the implementation, doesn't seem to difficult.

Reply:
Yes, thanks !